design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+  # routes can be present in any order, and will be matched from most to least
+  # specific.
+  routes:
+  # each route entry starts with a prefix match
+  # and one of service or delegate
+  - match: /static
+    # service defines the properties of the service in the current namespace
+    # that will handle traffic matching the route
+    service:
+    - name: google-static
+      port: 9000
+  - match: /finance
+    # delegate delegates the matching route to another IngressRoute object.
+    # This delegates the responsibility for /finance to the IngressRoute matching the delegate parameters
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
+  - match: /ads
+    # more than one service name/port pair may be present allowing the services for a route to
+    # be handled by more than one service.
+    service:
+    - name: ads-red
+      port: 9090
+    - name: ads-blue
+      port: 9090
+```
+
+This is an example of the `google-finance` object which has been delegated responsibilty for paths starting with `/finance`.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google-finance
+  namespace: finance
+spec:
+  # note that this is an edge, so there is no virtualhost key
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific, however as this is an edge, only prefixes that match the prefix
+  # that delegated to this object.
+  routes:
+  # each route in this object must start with /finance as this was the prefix which
+  # delegated to this object.
+  - match: /finance
+    service:
+    - name: finance-app
+      port: 9999
+  - match: /finance/static
+    service:
+    - name: finance-static-content
+      port: 8080
+  - match: /finance/partners
+    # delegate the /finance/partners prefix to the partners namespace
+    delegate:
+      name: finance-partners
+      namespace: partners
+```
+
+## Delegation rules
+
+The delegation rules applied are as follows
+
+1. If an `IngressRoute` object contains a `spec.virtualhost` key it is considered a root.
+2. If an `IngressRoute` object does not contain `spec.virtualhost` key is considered an edge.
+3. An edge is reachable if a delegation to it exists in another `IngressRoute` object.
+4. Edges which are not reachable are considered orphened. Orphened edges have no effect on the running configuration.
+
+## Validation rules
+
+The validation rules applied in this design are as follows
+Some of these rules may be relaxed in future designs.
+
+1. If a validation error is encountered, the entire object is rejected. Partial application of the valid portions of the configuration is _not_ attempted.
```

Comment:
How will validation errors be surfaced to the user? Maybe `validationStatus(es)` under `status`?

Reply:
Unknown. It could be on status, i could be in events.

Reply:
Events only last an hour, so need to be careful utilizing those fully.

Reply:
cool, thanks for explaining it. I think I need to do a followup with a detailed status: section, but I'd like to take that on notice for this review if you'll permit me.